reframe/frontend/autodetect.py

```diff
@@ -0,0 +1,153 @@
+# Copyright 2016-2021 Swiss National Supercomputing Centre (CSCS/ETH Zurich)
+# ReFrame Project Developers. See the top-level LICENSE file for details.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+
+import json
+import os
+import tempfile
+
+import reframe as rfm
+import reframe.utility.osext as osext
+from reframe.core.logging import getlogger
+from reframe.core.runtime import runtime
+from reframe.core.schedulers import Job
+from reframe.utility.cpuinfo import cpuinfo
+
+
+def _load_info(filename):
+    try:
+        with open(filename) as fp:
+            return json.load(fp)
+    except OSError as e:
+        getlogger().warning(
+            f'could not load file: {filename!r}: {e}'
+        )
+        return {}
+
+
+def _save_info(filename, topo_info):
+    if not topo_info:
+        return
+
+    os.makedirs(os.path.dirname(filename), exist_ok=True)
+    try:
+        with open(filename, 'w') as fp:
+            json.dump(topo_info, fp, indent=2)
+    except OSError as e:
+        getlogger().warning(
+            f'could not save topology file: {filename!r}: {e}'
+        )
+
+
+def _is_part_local(part):
+    return (part.scheduler.registered_name == 'local' and
+            part.launcher_type.registered_name == 'local')
+
+
+def _remote_detect(part):
+    getlogger().info(
+        f'Detecting topology of remote partition {part.fullname!r}'
+    )
+    rfm_exec = os.path.join(rfm.INSTALL_PREFIX, 'bin/reframe')
+    try:
+        with tempfile.TemporaryDirectory(dir='.') as dirname:
+            job = Job.create(part.scheduler,
+                             part.launcher_type(),
+                             name='rfm-detect-job',
+                             sched_access=part.access)
+            with osext.change_dir(dirname):
+                launcher_cmd = job.launcher.run_command(job)
+                job.prepare([f'{launcher_cmd} {rfm_exec} '
+                             f'--detect-host-topology=topo.json'],
+                            trap_errors=True)
```

Comment:
This assumes that the Python environment from the login node is also valid in the compute node, which is not necessarily true (I think). When I run this on Ault's `a64fx` partition, the cpu topology doesn't get detected because the above job gives me the classic `couldn't find module blah blah ...`. So I think this section here might need to pull the whole of ReFrame into the temp directory and there do the `sh bootstrap.sh`. This would make it more portable, I think.

Reply:
I just hacked the code to manually reinstall ReFrame in the temper from scratch and it worked like a charm. Since this will cache the output the first time around, I think that the extra overhead from the full reinstall is not a problem at all.

Reply:
Good point. When writing this, I assumed that the same environment was available on the compute nodes, but I didn't document it. In any case, I could update the generated script to reinstall reframe.

Reply:
How did you install it? I see three different ways, each on with its own advantages and disadvantages:

a. Using `pip` by doing something like `PYTHONPATH=external/ python3 -m pip install . --target=${tmp_path}`
b. Doing a fresh clone into a temp dir and bootstrap again
c. Copy the current installation to tempdir and re-bootstrap it.

Reply:
Also I guess that the temp dir is on the remote host, because we can't make the assumption that a temp dir on the local host exists on the remote one.

Reply:
I think I went for the option c to avoid dealing with the checking out the correct branch. In that case option b would just not work because my changes were only locally and I would clone a version of reframe which does not have the feature I was playing around with.

Reply:
Yes, I was thinking about (c) as well, but the downside is that if we copy the whole dir, we may also copy trash, that's why I was thinking about (a), which again has some other problems. But in any case, the TMPDIR you're copying into from the host may not be valid in the remote host.

Reply:
I mean, once implemented, b should always work. The tricky bit is how to test it while developing it 😅

Reply:
I like better the solution that makes use of `cp (option `c`), rather than `a` and `b`.
`a` has the problem that @jjotero has mentioned and `b` may be a different version than the one that I am running, which can cause problems if the file format or fields change between ReFrame versions.
Copying is more consistent. But I am not sure how one can bootstrap after copying if the compute nodes do not have `pip` or no internet access, which is a problem for all solutions...

Reply:
I am personally a bit reluctant with copy, because you don't know what you are going to copy. We will be copying `external/` which can be hundreds of MB and even some stale stage directories which could be several GB.

Reply:
I agree. I guess we could try/except the current option, then b and then c?

Reply:
I wouldn't go for (c), because that would mean additional configuration. If we cannot run the remote command, we will issue a warning and the user can find a way to manually run `--detect-host-topology` and copy the metaconfig file.

Reply:
So, current solution + a fresh clone with bootstrap as a fallback.

Reply:
Also, I'm not passing the `FILE` argument to the `--detect-host-topology` anymore, because that would imply that we need to have a common shared directory between the local and the remote hosts. I'm just using the standard output and if it's not a JSON (case of errors), I'm issuing a warning and continue with the next option to try.

Reply:
This is not resolved yet. I'm still working on it...

Reply:
Finally, this is done 🎉  Actually, I'm taking the option (b) because it has the less side effects if you do it in a controlled way, as I do. Actually, "installing" reframe anywhere is pretty easy: 1. Copy `['bin/', 'reframe/', 'bootstrap.sh', 'requirements.txt']` to the target directory, 2. Run `./bootstrap.sh` from that directory. So this is what I'm doing. The `./bootstrap.sh` is always run from the generated job script, so it covers nicely the Ault case. By default, I'm copying to a temporary directory under `.`, assuming this directory is most likely shared with the compute nodes (I want to avoid `$TMPDIR` since most likely -if not always- is not shared). But if you want to run from anywhere else than `.`, you can set the prefix to `RFM_REMOTE_WORKDIR`. Let me know what do you think now.